File: androidApp/src/main/java/com/alextos/darts/android/common/presentation/views/GameRecapView.kt
package com.alextos.darts.android.common.presentation.views

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.alextos.darts.android.R
import com.alextos.darts.android.common.presentation.components.ChartLegendItem
import com.alextos.darts.android.common.presentation.components.SectionHeader
import com.alextos.darts.core.domain.model.Player
import com.alextos.darts.core.util.numberOfTurns
import com.alextos.darts.game.domain.models.GameDuration
import com.alextos.darts.game.domain.models.PlayerGameValue
import com.alextos.darts.game.domain.models.PlayerHistory
import com.github.tehras.charts.line.LineChart
import com.github.tehras.charts.line.LineChartData
import com.github.tehras.charts.line.renderer.line.SolidLineDrawer
import com.github.tehras.charts.line.renderer.point.FilledCircularPointDrawer
import com.github.tehras.charts.piechart.animation.simpleChartAnimation

private val colors = listOf(Color.Red, Color.Green, Color.Blue, Color.Yellow, Color.Magenta, Color.Cyan)

@Composable
fun GameRecapView(
    history: List<PlayerHistory>,
    averageTurns: List<PlayerGameValue>,
    biggestTurns: List<PlayerGameValue>,
    smallestTurns: List<PlayerGameValue>,
    misses: List<PlayerGameValue>,
    overkills: List<PlayerGameValue>,
    duration: GameDuration,
    numberOfTurns: Int
) {
    if (history.isNotEmpty()) {
        LazyColumn(
            modifier = Modifier
                .fillMaxSize(),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            item {
                RecapLineChart(history = history)
            }
            item {
                RecapLineChartLegend(players = history.map { it.player })
            }
            if (!duration.isEmpty) {
                item {
                    GameDuration(duration = duration)
                }
            }

            item {
                NumberOfTurns(numberOfTurns = numberOfTurns)
            }

            item {
                ValuesBlock(values = averageTurns, title = stringResource(id = R.string.average_set_recap))
            }
            item {
                ValuesBlock(values = biggestTurns, title = stringResource(id = R.string.biggest_sets))
            }
            item {
                ValuesBlock(values = smallestTurns, title = stringResource(id = R.string.smallest_sets))
            }
            item {
                ValuesBlock(values = misses, title = stringResource(id = R.string.misses_count))
            }
            item {
                ValuesBlock(values = overkills, title = stringResource(id = R.string.overkill_count))
            }
            item {
                Spacer(modifier = Modifier.height(72.dp))
            }
        }
    }
}

@Composable
private fun ValuesBlock(values: List<PlayerGameValue>, title: String) {
    Column(modifier = Modifier
        .fillMaxWidth()
        .padding(top = 32.dp)
    ) {
        SectionHeader(title = title)
        values.forEach { set ->
            Item(name = set.player.name, value = set.value)
        }
    }
}

@Composable
private fun Item(name: String, value: Int) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 4.dp),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(name)
        Text(value.toString())
    }
}

@Composable
private fun RecapLineChart(history: List<PlayerHistory>) {
    val maxSize = history.numberOfTurns()
    val data = history.map { playerHistory ->
        LineChartData(
            points = playerHistory.chartData(maxSize).map { value ->
                LineChartData.Point(value, value.toString())
            },
            lineDrawer = SolidLineDrawer(
                color = colors.getOrNull(history.indexOf(playerHistory)) ?: Color.DarkGray
            ),
            startAtZero = true
        )
    }
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .aspectRatio(1f)
            .background(Color.White),
        contentAlignment = Alignment.Center
    ) {
        LineChart(
            linesChartData = data,
            modifier = Modifier
                .fillMaxSize(0.8f)
                .aspectRatio(1f),
            pointDrawer = FilledCircularPointDrawer(),
            horizontalOffset = 5f,
            labels = listOf(),
            animation = simpleChartAnimation()
        )
    }
}

@Composable
private fun RecapLineChartLegend(players: List<Player>) {
    Column(modifier = Modifier.padding(horizontal = 16.dp)) {
        players.forEach {
            val color = colors.getOrNull(players.indexOf(it)) ?: Color.DarkGray
            ChartLegendItem(text = it.name, color = color)
        }
    }
}

@Composable
private fun GameDuration(duration: GameDuration) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp),
        horizontalArrangement = Arrangement.End
    ) {
        Text(stringResource(id = R.string.game_duration, duration.minutes, duration.seconds))
    }
}

@Composable
private fun NumberOfTurns(numberOfTurns: Int) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp),
        horizontalArrangement = Arrangement.End
    ) {
        Text(stringResource(id = R.string.number_of_turns, numberOfTurns))
    }
}

File: iosApp/iosApp/Library/Views/Views/GameRecapView.swift
import SwiftUI
import shared
import Charts

struct GameRecapView: View {
	let history: [PlayerHistory]
	let averageTurns: [PlayerGameValue]
	let biggestTurns: [PlayerGameValue]
	let smallestTurns: [PlayerGameValue]
	let misses: [PlayerGameValue]
	let overkills: [PlayerGameValue]
	let duration: GameDuration
	let numberOfTurns: Int32
	
	@Environment(\.displayScale) var displayScale
	@State private var renderedImage = Image(systemName: "photo")
	
	var body: some View {
		ScrollView {
			content
		}
		.background(Color.background)
		.toolbar {
			ToolbarItem(placement: .navigationBarTrailing) {
				ShareLink("share", item: renderedImage, preview: SharePreview("preview", image: renderedImage))
			}
		}
		.task {
			do {
				try await Task.sleep(nanoseconds: 1_000_000_000)
				render()
			} catch {}
		}
	}
	
	@ViewBuilder
	private var content: some View {
		VStack {
			chart
			if !duration.isEmpty {
				gameDurationItem
			}
			numberOfTurnsItem
			valuesBlock(values: averageTurns, title: String(localized: "average_set_recap"))
			valuesBlock(values: biggestTurns, title: String(localized: "biggest_sets"))
			valuesBlock(values: smallestTurns, title: String(localized: "smallest_sets"))
			valuesBlock(values: misses, title: String(localized: "misses_count"))
			valuesBlock(values: overkills, title: String(localized: "overkill_count"))
		}
	}
	
	@ViewBuilder
	private var chart: some View {
		Chart {
			let maxSize = history.max(by: { $0.turns.count < $1.turns.count })?.turns.count ?? 0
			
			ForEach(history, id: \.id) { playerHistory in
				
				let data = playerHistory.chartData(size: Int32(maxSize))
				ForEach(data.indices, id: \.self) { index in
					LineMark(
						x: .value("", index),
						y: .value("", data[index].floatValue)
					)
				}
				.foregroundStyle(by: .value("Player", playerHistory.id))
				.symbol(by: .value("Player", playerHistory.id))
			}
		}
		.aspectRatio(1.0, contentMode: .fit)
		.padding()
	}
	
	@ViewBuilder
	private var gameDurationItem: some View {
		HStack {
			Spacer()
			Text("game_duration \(duration.minutes) \(duration.seconds)")
				.font(.caption)
		}
		.padding(.horizontal, 16)
	}
	
	@ViewBuilder
	private var numberOfTurnsItem: some View {
		HStack {
			Spacer()
			Text("number_of_turns \(numberOfTurns)")
				.font(.caption)
		}
		.padding(.horizontal, 16)
	}
	
	@ViewBuilder
	private func valuesBlock(values: [PlayerGameValue], title: String) -> some View {
		VStack {
			SectionHeader(title: title)
			ForEach(values) { value in
				item(name: value.player.name, value: value.value)
			}
		}
		.padding(.top, 32)
	}
	
	@ViewBuilder
	private func item(name: String, value: Int32) -> some View {
		HStack {
			Text(name)
			Spacer()
			Text("\(value)")
		}
		.padding(.horizontal, 16)
		.padding(.vertical, 4)
	}
	
	@MainActor
	private func render() {
		let content = content.background(Color.background)
			.frame(width: 300)
		let renderer = ImageRenderer(content: content)
		// make sure and use the correct display scale for this device
		renderer.scale = displayScale
		if let uiImage = renderer.uiImage {
			renderedImage = Image(uiImage: uiImage)
		}
	}
}
